count_ops and matching related changes

smichr opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
fixes #15907
fixes #17498

#### Brief description of what is fixed or changed
* match ignores bound symbols unless requested
If you are trying to match an object having bound symbols, Wild placeholders no longer need to be used to allow for differences in bound symbols, e.g.
```python
>>> W1,W2=map(Wild, 'ab')
>>> Sum(y, (y, 1, 3)).match(Sum(x, (x, W1, W2)))
{b_: 3, a_: 1}
```
* as_dummy copies symbol commutativity
It is explicitly stated that `as_dummy` will return a unique symbol for a symbol (vanilla except for matching commutativity) or expression with with canonical symbols for the bound symbols.

* make canonical_variables more canonical
Even when clashing symbols appear in the bound symbols, these will not prevent the canonical representation of an object with bound symbols
```python
>>> _0=Symbol("_0")
>>> Sum(x + _0, (x, 1, 2), (_0, 2, 3)).as_dummy()
Sum(_0 + _1, (_0, 1, 2), (_1, 2, 3))
```
Only clashing free symbols will force the bound symbol numbering to change
```python
>>> Sum(x + y, (x, _0, 2), (y, 2, 3)).as_dummy()
Sum(_1 + _2, (_1, _0, 2), (_2, 2, 3))
```

* optimizations
  - use xreplace in dummy_eq
  - make replace more robust; the mechanism of replacing matched objects with a Dummy was fragile in that it did not allow for the fact that some objects in a tree cannot be represented by a symbol.
  - use == not is for comparison; I happened across the use of something equivalent to `x is not min(x, y)` which was changed to `x != min(x, y)` because `x` and `y` were not guaranteed to be singletons.
  - the match dictionary should no longer have extraneous (unrequested) wild symbols in it anymore
* count_ops

`Sum`, unlike `Integral`, did not show up in `count_ops` results. After reviewing the code I tidied it up a bit and included `Sum`.

before:
```python
>>> Sum(x, (x, 1, 2)).count_ops()
0
```
now:
```python
>>> Sum(x, (x, 1, 2)).count_ops(visual=1)
SUM
```

* matching

The use of `ordered` while matching sometimes orders terms of Add in a way that leads to sub-optimal matching:

before:
```python
>>> X, Y = map(Wild, "XY")
>>> x, y = symbols('x y')
>>> (5*y - x).match(5*X - Y)
{Y_: -5*y, X_: -x/5}
```
now:
```python
>>> (5*y - x).match(5*X - Y)
{X: y, Y: x}
```

* Lambda comparison

A custom `__eq__` routine was removed to prevent Lambdas (that are functionally the same but written in different variables) from comparing equal. Good discussions are reasons are referenced or given in #5593.

#### Other comments

Although I tried several modifications of `dummy_eq` (in response to #19266) I am leaving it alone. It should probably be as @oscarbenjamin pointed out below. Here, too, is a method comparing to expression to see if they are the same except for symbol names:
```python
    def dummy_eq(self, other, symbol=None):
        """
        Return True if the expression is the same except for
        possible different use of bound symbols, else False.

        Examples
        ========

        >>> from sympy import Dummy, Sum, symbols
        >>> from sympy.abc import x, y, z
        >>> _d, _e, _f = symbols('d e f', cls=Dummy)

        Normal equality tests fail if the symbols are not
        the same:

        >>> _d == _e
        False

        This even applies for objects that use bound symbols
        (and it is important for the Python inner-workings of
        SymPy that this be so):

        >>> Sum(x, (x, 1, 2)) == Sum(y, (y, 1, 2))
        False

        The purpose of ``dummy_eq`` is to answer whether two
        expressions are the same except for bound symbol names:

        >>> Sum(x, (x, 1, 2)).dummy_eq(Sum(y, (y, 1, 2)))
        True

        Any difference in structure or names of free symbols will
        result in False.

        """
        return self.as_dummy() == _sympify(other).as_dummy()

    def symap(self, other, map=False):
        """return True if ``self`` and ``other`` are the same except
        for symbol names, else False. If ``map`` is True then return
        None instead of False and the mapping between free symbols of
        ``self`` and ``other``.

        Examples
        ========

        >>> from sympy.abc import x, y
        >>> x.symap(x), x.symap(x, True)
        (True, {})
        >>> x.symap(y), x.symap(y, True)
        (True, {x: y})
        >>> x.symap(x + 1), x.symap(x + 1, True)
        (None, False)

        """
        from sympy import Wild, Symbol
        other = _sympify(other)
        s, o = [i if i.is_Symbol else i.as_dummy() for i in (self, other)]
        reps = {i: Wild(i.name) for i in s.free_symbols}
        pat = s.xreplace(reps)
        m = other.match(pat)
        if m is not None and all(i.is_Symbol for i in m.values()):
            if not map:
                return True
            ireps = {v: k for k, v in reps.items()}
            return {ireps[w]: v for w, v in m.items() if ireps[w] != v}
        return False if map else None
```
#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* core
    * `Sum` is now included in `count_ops` results
    * `match` has been optimized to improve term-matching in Add
    * `match` will automatically ignore bound symbols when matching
    * `match` results will no longer contain extraneous symbols (only those requested)
    * `as_dummy` should give a canonical result for expressions having bound symbols
    * Lambdas written in terms of different symbols will no longer compare equal; this was an abuse of the Python `==` operator
* sets
    * ImageSets with Lambdas having different symbols will no longer compare equal and should be compared like `a.dummy_eq(b)`
<!-- END RELEASE NOTES -->

smichr committed:
Sum shows in count_ops

oscarbenjamin commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1143,12 +1143,21 @@ def test_Sum_dummy_eq():
     assert Sum(x, (x, a, d)).dummy_eq(Sum(x, (x, a, c)), c)
-    assert not Sum(x, (x, a, d)).dummy_eq(Sum(x, (x, a, c)))
-    assert Sum(x, (x, a, c)).dummy_eq(Sum(y, (y, a, c)))
+    assert Sum(x, (x, a, d)).dummy_eq(Sum(x, (x, a, c)))
```
This doesn't seem right. Surely only x is a dummy variable here. If c and d are unequal then these two sums are unequal.

smichr replied:
There are two types of dummy symbols: structurally bound (e.g. the summation variable) and physical Dummy symbols. `Expr.as_dummy` will make two expressions that differ only in bound symbols the same. `Expr.dummy_eq` is more of a special form of matching and answers the question of whether two expressions are the same except for names of symbols. And by default it limits attention to the Dummy symbols of self.

In this case, d is the Dummy and the two expression are the same if `d -> c` so that's why `dummy_eq` reports True.

Often this routine can be used in testing. In your PR I noticed that `trigintegrate(cos(x)*cot(x), x)` gives something equivalent to
```python
>>> eq
(-_b1 + _b2)*log(cos(x) + 1)/2 + (_b1 + _b2)*log(cos(x) - 1)/2 + cos(x) + 1
```
where the symbols with leading underscores are Dummy symbols. To test this in this PR you can do
```python
_b1, _b2 = Dummy(), Dummy()  # note that these are now different than those in eq
ans = (-_b1 + _b2)*log(cos(x) + 1)/2 + (_b1 + _b2)*log(cos(x) - 1)/2 + cos(x) + 1
assert eq.dummy_eq(ans)
```
Though I suspect you know this, I am reiterating for the sake of others: the type and names used in the arg to `dummy_eq` are immaterial (provided that they do not clash with a symbol already in the expression) so the following would also pass:
```python
b1, b2 = symbols('b1 b2')
ans = (-b1 + b2)*log(cos(x) + 1)/2 + (b1 + b2)*log(cos(x) - 1)/2 + cos(x) + 1
assert eq.dummy_eq(ans)
```

oscarbenjamin replied:
To me that seems like changing the definition of `dummy_eq`.

For testing what we need is just a function that compares two expressions as equal under a renaming of the Dummy symbols. I wouldn't want to use this version of `dummy_eq` in testing if it gives things like
```julia
In [2]: Symbol('x').dummy_eq(Symbol('y'))                                                                                                      
Out[2]: True
```

smichr replied:
Hopefully you are not getting that with this branch:

```python
>>> Symbol('x').dummy_eq(Symbol('y'))
False
```

The default behavior of the method is to do as you said: see if Dummy variables in self can be uniquely mapped to symbols in other to give two expressions that are the same. The behavior can be slightly adapted, however, by providing `symbol` in which case the mapping goes from the specified symbols in `other` to `self`.

oscarbenjamin replied:
I thought it was this branch but can't reproduce it now. Maybe it was before the force-push...

Okay I see what you mean. I hadn't noticed that `d` was a `Dummy` above. Maybe rename it to `_d`?

smichr replied:
I do that in the doctest which are more public-facing. In the test suite `d = Dummy()` appears just before the test so I think it's ok.

smichr replied:
> Maybe it was before the force-push...

I didn't think of that til later -- yes, I corrected that condition and added coverage apparently after you had pulled a version.

smichr commented on sympy/core/basic.py:
```diff
@@ -1628,8 +1746,21 @@ def match(self, pattern, old=False):
         """
-        pattern = sympify(pattern)
-        return pattern.matches(self, old=old)
+        from sympy.core.symbol import Wild
```
```suggestion
        from sympy.core.symbol import Wild, Symbol
```

smichr commented on sympy/core/basic.py:
```diff
@@ -1618,6 +1731,11 @@ def match(self, pattern, old=False):
+        Structurally bound symbols are ignored during matching:
+
+        >>> Sum(x, (x, 1, 2)).match(Sum(y, (y, 1, p)))
+        {_p: 2}
```
```suggestion
        {p_: 2}
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,89 @@ def __ne__(self, other):
-                return s == o
+        This is done automatically for univarite expressions:
+
+        >>> (1 + _d).dummy_eq(1 + x)
```
```suggestion
        >>> (1 + _d).dummy_eq(1 + x)
        True
```

smichr commented on sympy/core/tests/test_equal.py:
```diff
@@ -75,13 +75,25 @@ def test_cmp_issue_4357():
+
+    assert x.dummy_eq(x)
+    assert x.dummy_eq(x, x)
+    assert x.dummy_eq(x, True)
```
```suggestion
```

smichr commented on sympy/core/tests/test_equal.py:
```diff
@@ -75,13 +75,25 @@ def test_cmp_issue_4357():
+    assert x.dummy_eq(x, True)
+    assert not (1 + _v).dummy_eq(x)
+    assert not x.dummy_eq(y)
+    assert not _u.dummy_eq(y)
```
```suggestion
    assert _u.dummy_eq(y)
```

smichr commented on sympy/core/tests/test_equal.py:
```diff
@@ -75,13 +75,25 @@ def test_cmp_issue_4357():
+    assert not _u.dummy_eq(y, (x, _v))
+    assert _u.dummy_eq(y, (y, _v))
+
+    assert (_u**2 + 1).dummy_eq(x**2 + 1) is False, (_u**2 + 1).dummy_eq(x**2 + 1)
```
```suggestion
    assert (_u**2 + 1).dummy_eq(x**2 + 1)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,90 @@ def __ne__(self, other):
+        if m is None:
+            return False
+        # they must have matched all free Dummy symbols in self
+        return not set(dfree) - set(m.values())
```
```suggestion
        return not set(dfree) - set(m.values()) and s == o.xreplace(m)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,94 @@ def __ne__(self, other):
+        >>> (1 + _d).dummy_eq(1 + x)
+        True
+        """
+        from sympy.core.symbol import Wild, Symbol
```
```suggestion
        from sympy.core.symbol import Dummy, Symbol, Wild
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,94 @@ def __ne__(self, other):
+                u = ofree - sfree
+                if len(u) == 1:
+                    u = u.pop()
+                    return s.dummy_eq(o.xreplace({u: Dummy(u.name)}), u)
```
```suggestion
                    d = Dummy()
                    return s.dummy_eq(o.xreplace({u: d}), d)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,95 @@ def __ne__(self, other):
+
+        if symbol is None:
+            ofree = o.free_symbols
+            if len(sfree) != len(ofree):
```
```suggestion
            if len(dfree) != len(ofree):
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,95 @@ def __ne__(self, other):
+            symbol = [i for i in ofree if i.is_Dummy]
+            if not symbol:
+                u = ofree - sfree
+                if len(u) == 1:
```
```suggestion
                if len(u) == len(dfree) == 1:
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,95 @@ def __ne__(self, other):
+        if symbol is None:
+            ofree = o.free_symbols
+            if len(sfree) != len(ofree):
+                return False
```
```suggestion
                print(1,'so',sfree,ofree)
                return False
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,95 @@ def __ne__(self, other):
+                if len(u) == 1:
+                    u = u.pop()
+                    d = Dummy()
+                    return s.dummy_eq(o.xreplace({u: d}), d)
```
```suggestion
                    print(s,o.xreplace({u: d}), d.s.dummy_eq(o.xreplace({u: d}), d))
                    return s.dummy_eq(o.xreplace({u: d}), d)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,95 @@ def __ne__(self, other):
+                    u = u.pop()
+                    d = Dummy()
+                    return s.dummy_eq(o.xreplace({u: d}), d)
+                return False
```
```suggestion
                print(u,dfree)
                return False
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,95 @@ def __ne__(self, other):
+        s, o = [i if i.is_Symbol else i.as_dummy() for i in (s, o)]
+        if s == o:
+            return True
+        sfree = self.free_symbols
```
```suggestion
        sfree = s.free_symbols
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,99 @@ def __ne__(self, other):
+            return True
+        sfree = s.free_symbols
+        dfree = [i for i in sfree if i.is_Dummy]
+        if not dfree and symbol != True:
```
```suggestion
        if not (dfree or symbol):
```

smichr commented on sympy/core/tests/test_equal.py:
```diff
@@ -75,13 +75,33 @@ def test_cmp_issue_4357():
+    assert not y.dummy_eq(_u)
+    assert not _u.dummy_eq(y, x)
+    assert not _u.dummy_eq(y, (x, _v))
+    # univariate rhs auto-supplies self
+    assert _u.dummy_eq(y)
+    assert (_u**2 + 1).dummy_eq(x**2 + 1)
```
```suggestion
    # univariate rhs no longer auto-supplies symbol
    assert not _u.dummy_eq(y)
```

smichr commented on sympy/core/tests/test_equal.py:
```diff
@@ -75,13 +75,33 @@ def test_cmp_issue_4357():
+    # univariate rhs auto-supplies self
+    assert _u.dummy_eq(y)
+    assert (_u**2 + 1).dummy_eq(x**2 + 1)
+    # but can be specified
+    assert _u.dummy_eq(y, (y, _v))
+    # and can be disabled with `strict`
+    assert not (_u**2 + 1).dummy_eq(x**2 + 1, strict=True)
```
```suggestion
    # but can be specified
    assert _u.dummy_eq(y, (y, _v))  # ignore extras
```

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1139,16 +1139,24 @@ def test_issue_15943():
     assert Sum(x, (x, a, c)).dummy_eq(Sum(y, (y, a, c)))
-    assert Sum(x, (x, a, d)).dummy_eq(Sum(y, (y, a, c)), c)
-    assert not Sum(x, (x, a, d)).dummy_eq(Sum(y, (y, a, c)))
+    # not Dummy on rhs
```
```suggestion
    # no Dummy on rhs
```

smichr commented on sympy/concrete/tests/test_sums_products.py:
```diff
@@ -1139,16 +1139,24 @@ def test_issue_15943():
+    assert Sum(x, (x, a, _d)).dummy_eq(Sum(y, (y, a, c)), c)
+    assert not Sum(x, (x, a, _d)).dummy_eq(Sum(x, (x, b, c)), c)
+    # no Dummy on left
+    assert Sum(x, (x, a, c)).dummy_eq(Sum(y, (y, a, _d)), _d)
+    assert not Sum(x, (x, a, c)).dummy_eq(Sum(y, (y, a, _d)))
```
```suggestion
    assert not Sum(x, (x, a, c)).dummy_eq(Sum(y, (y, a, _d)), _d)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,99 @@ def __ne__(self, other):
+        sfree = s.free_symbols
+        dfree = [i for i in sfree if i.is_Dummy]
+        if not (dfree or symbol):
+            return False
```
```suggestion
            return False
        strict = symbol != True
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,99 @@ def __ne__(self, other):
+        w = o.xreplace({i: Wild(i.name) for i in symbol})
+        m = s.match(w) or {}
+        # they must have matched all free Dummy symbols in self
+        found_all = not set(dfree) - set(m.values())
```
```suggestion
        found_all = not set(dfree) - set(m.values())
        if strict and not all(i.is_Dummy for i in m.values())
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,99 @@ def __ne__(self, other):
+            else:
+                if len(sfree) != len(ofree):
+                    return False
+                symbol = [i for i in ofree if i.is_Dummy]
```
```suggestion
                symbol = [i for i in ofree if i.is_Dummy]
                if strict and len(symbol) > len(dfree):
                    return False  # won't be able to match all
```

smichr commented on sympy/core/tests/test_equal.py:
```diff
@@ -75,13 +75,30 @@ def test_cmp_issue_4357():
+    assert not (x + 1).dummy_eq(y + 1, _u)
+    assert not (1/x + y).dummy_eq(x + _u, _u)
+    assert not (_u + _v).dummy_eq(x)
+    assert not (_u + _v).dummy_eq(_u + x)
```
```suggestion
    # not all Dummy on left have a Dummy on right
    assert not (_u + _v).dummy_eq(_u + x)
    # in this case, if x is made a Dummy then x in self
    # no longer matches
    assert not (_u + x).dummy_eq(_v + x, x)
    # not all implicit Dummy on right have a Dummy to match on left
    assert not (_u + x).dummy_eq(_v + x, True)
    assert not Dummy().dummy_eq(_u + _v)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,103 @@ def __ne__(self, other):
+        m = s.match(w) or {}
+        # they must have matched all free Dummy symbols in self
+        found_all = not set(dfree) - set(m.values())
+        if strict and not all(i.is_Dummy for i in m.values())
```
```suggestion
        if strict and not all(i.is_Dummy for i in m.values()):
            return False
```

smichr commented on sympy/core/tests/test_equal.py:
```diff
@@ -75,13 +75,37 @@ def test_cmp_issue_4357():
+    # no longer matches
+    assert not (_u + x).dummy_eq(_v + x, x)
+    # not all implicit Dummy on right have a Dummy to match on left
+    assert not (_u + x).dummy_eq(_v + x, True)
+    assert not Dummy().dummy_eq(_u + _v)
```
```suggestion
    assert not Dummy().dummy_eq(_u + _v)
    # symbol=True ignores identity and asserts structural equality
    # via a 1:1 symbol mapping
    assert (_u + x).dummy_eq(_v + x, True)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,104 @@ def __ne__(self, other):
+        found_all = not set(dfree) - set(m.values())
+        if strict and not all(i.is_Dummy for i in m.values()):
+            return False
+        if not found_all:
```
```suggestion
        if not strict and not all(i.is_Symbol for i in m.values()):
            return False
        if not found_all:
```

smichr commented on sympy/core/tests/test_equal.py:
```diff
@@ -75,13 +75,41 @@ def test_cmp_issue_4357():
+    # via a 1:1 symbol mapping
+    assert (_u + x).dummy_eq(_v + x, True)
+    # and it checks that a symbol:symbol match has been made
+    assert (1 + _u).dummy_eq(x + y, True)
```
```suggestion
    assert not (1 + _u).dummy_eq(x + y, True)
```

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,106 @@ def __ne__(self, other):
+
+        If there are no Dummy symbols in ``self`` the result
+        can only be true if the expressions are structurally the
+        same in their use of free symbols.
```
To me this seems like a separate function from `dummy_eq` that should have a different name.

oscarbenjamin commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,106 @@ def __ne__(self, other):
+        >>> a = x + 2*y
+        >>> b = y + 2*x
+        >>> a.dummy_eq(b, True)
+        True
```
To me this seems like a separate function from dummy_eq that should have a different name.

smichr replied:
Would you be in favor of putting `dummy_eq` on a deprecation path and replacing it with `sym_eq` which will tell if `self` and `other` are symbolically the same in specified symbols, defaulting to all?

```python
x.sym_eq(y)
Sum(x, (x, 1, 2)).sym_eq(Sum(y, (y, 1, 2)))
(x + y).sym_eq(x + z, z)
(Dummy() + y).sym_eq(x + y)
(Dummy() + y).sym_eq(x + y, x) # y->y, x->Dummy()
not (Dummy() + y).sym_eq(x + y, y)  # nothing for x to match to
```

oscarbenjamin replied:
I'm not sure I understand what you propose.

What is the reason for wanting to change `dummy_eq`?

smichr replied:
OK, I made `dummy_eq` to return true for only 3 cases:

1) _aresame(self, other)
2) Dummy in self maps to Dummy in other (default)
3) Dummy in self maps to indicated symbols in other

smichr replied:
This breaks the hopefully-not-used `x.dummy_eq(Dummy())` syntax.

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,116 @@ def __ne__(self, other):
+        from sympy import Wild
+        other = _sympify(other)
+        sfree = self.free_symbols
+        pat = self.xreplace({i: Wild(i.name)} for i in symbols})
```
```suggestion
        s, o = [i if i.is_Symbol else i.as_dummy() for i in (self, other)]
        reps = {i: Wild(i.name) for i in s.free_symbols}
        pat = self.xreplace(reps)
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,116 @@ def __ne__(self, other):
+        pat = self.xreplace({i: Wild(i.name)} for i in symbols})
+        m = other.match(pat)
+        if m is not None and all(i.is_Symbol for i in m.values()):
+            return {k: v for k, v in m.items() if k != v}
```
```suggestion
            ireps = {v: k for k, v in reps.items()}
            return {ireps[w]: v for w, v in m.items() if ireps[w] != v}
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,116 @@ def __ne__(self, other):
+        m = other.match(pat)
+        if m is not None and all(i.is_Symbol for i in m.values()):
+            return {k: v for k, v in m.items() if k != v}
+        return False
```
```suggestion
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,116 @@ def __ne__(self, other):
-            return s == o
+    def symap(self, other):
+        """return the symbol mapping if ``self`` and ``other`` are the same
+        except for the remapping of symbols else False.
```
```suggestion
        except for the remapping of symbols else None.
```

smichr commented on sympy/core/basic.py:
```diff
@@ -367,49 +367,116 @@ def __ne__(self, other):
+        {}
+        >>> x.symap(y)
+        {x: y}
+        >>> x.symap(x + 1)
+        False
```
```suggestion
        >>> x.symap(x + 1) is None
        True
```

oscarbenjamin commented:
I still don't understand why `dummy_eq` needs to be changed at all...

smichr commented:
> why `dummy_eq` needs to be changed at all.

* It can't test more than 1 dummy symbol -- tests in your domain solver generate 2 Dummies (`trigintegrate`). How will those be tested?

oscarbenjamin commented:
> * It can't test more than 1 dummy symbol

Okay I see that now.

Maybe that part of `dummy_eq` is just a bad idea anyway though. Really if we want the Dummies to compare equal is that not because they are in fact bound?

oscarbenjamin commented:
Are there any *good* examples using `dummy_eq` with a non-bound `Dummy`?

oscarbenjamin commented:
I tried this diff, ran the tests, and saw no particularly compelling failures:
```diff
diff --git a/sympy/core/basic.py b/sympy/core/basic.py
index bbb285a72f..8dc8aa9fde 100644
--- a/sympy/core/basic.py
+++ b/sympy/core/basic.py
@@ -388,28 +388,7 @@ def dummy_eq(self, other, symbol=None):
         False
 
         """
-        s = self.as_dummy()
-        o = _sympify(other)
-        o = o.as_dummy()
-
-        dummy_symbols = [i for i in s.free_symbols if i.is_Dummy]
-
-        if len(dummy_symbols) == 1:
-            dummy = dummy_symbols.pop()
-        else:
-            return s == o
-
-        if symbol is None:
-            symbols = o.free_symbols
-
-            if len(symbols) == 1:
-                symbol = symbols.pop()
-            else:
-                return s == o
-
-        tmp = dummy.__class__()
-
-        return s.subs(dummy, tmp) == o.subs(symbol, tmp)
+        return self.as_dummy() == _sympify(other).as_dummy()
 
     # Note, we always use the default ordering (lex) in __str__ and __repr__,
     # regardless of the global setting.  See issue 5487.
```

smichr commented:
> Are there any _good_ examples using `dummy_eq` with a non-bound `Dummy`?

Searching through codebase:
* comparing Lambdas - obsolete since Lambda(x,x+1)==Lambda(y,y+1) now
* Polys that need to create a generator, e.g. `gegenbauer_poly(1, S.Half).dummy_eq(x)`

All the rest appear (looking quickly) to be bound-symbol related.

What about the 2 constants of integration that were being returned by your branch? Should we always be looking for a way to send back expressions with uniquely named symbols or should just use Dummy symbols?

oscarbenjamin commented:
> * comparing Lambdas - obsolete since `Lambda(x,x+1)==Lambda(y,y+1)` now

That should be changed. Either `Lambda` should canonicalise the symbols or those should compare unequal.

> * Polys that need to create a generator, e.g. `gegenbauer_poly(1, S.Half).dummy_eq(x)`

Maybe that API should be fixed to require a generator or return a `PurePoly`.

> What about the 2 constants of integration that were being returned by your branch?

I'm not sure which branch you are referring to.

smichr commented:
> I'm not sure which branch you are referring to.

#15396, e.g.:
```python
>>> from sympy.integrals.trigonometry import *
>>> trigintegrate(cos(x)*cot(x), x)
-(_b1 - _b2)*log(cos(x) + 1)/2 + (_b1 + _b2)*log(cos(x) + (_b2 - (_b1 + _b2))/_b1)/2 + cos(x) + 1
```


> That should be changed.

#17498

Lots of discussion at #15240 and your conclusions stated well [here](https://github.com/sympy/sympy/issues/5539#issuecomment-524504574).

smichr commented:
This is a bit of a mystery: how can a doctest pass and the identical test in the test suite fail?
```
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/testing/runtests.py", line 1335, in _timeout
    function()
  File "/home/travis/virtualenv/python3.8.0/lib/python3.8/site-packages/sympy-1.7.dev0-py3.8.egg/sympy/physics/quantum/tests/test_cg.py", line 150, in test_cg_simp_sum
    assert cg_simp(x * Sum(CG(a, alpha, b, 0, a, alpha), (alpha, -a, a)
AssertionError
```
vs this passing doctest 
```
    """
    >>> from sympy import symbols, Sum, KroneckerDelta
    >>> from sympy.physics.quantum.cg import CG, cg_simp
    >>> x, a, b, alpha = symbols('x a b alpha')
    >>> k = cg_simp(x * Sum(CG(a, alpha, b, 0, a, alpha), (alpha, -a, a)))
    >>> x*(2*a + 1)*KroneckerDelta(0, b)
    x*(2*a + 1)*KroneckerDelta(0, b)
    >>> _ == k
    True
```
Anyone have any ideas?

oscarbenjamin commented:
> Anyone have any ideas? I reverted the `count_ops` change and the problem still remains, so it has to do with the ordering of args during matching.

No idea. Ideally the matching should be done so that the ordering of the args doesn't affect the result...

smichr commented:
It is system/version dependent because it passes locally under 3.7 with x86 Windows.

smichr committed:
use xreplace in dummy_eq

xreplace can be used without issue since
bound symbols have already been unified

smichr committed:
make canonical_variables more canonical

smichr committed:
as_dummy copies symbol commutativity

smichr committed:
anticipate Lambda(x,x+1)!=Lambda(y,y+1)

oscarbenjamin commented on sympy/core/tests/test_function.py:
```diff
@@ -239,7 +238,9 @@ def test_Lambda():
 
-    assert Lambda(x, 2*x) + Lambda(y, 2*y) == 2*Lambda(x, 2*x)
+    eq = Lambda(x, 2*x) + Lambda(y, 2*y)
+    assert eq != 2*Lambda(x, 2*x)
```
Looking at this I suddenly realise the `Lambda` is a subclass of `Expr`. That should be changed...

smichr replied:
Why? Sum is an Expr, too, and 
```python
>>> eq = Sum(x, (x, 1, 2))
>>> eq + eq.xreplace({x:y})
Sum(x, (x, 1, 2)) + Sum(y, (y, 1, 2))
```

oscarbenjamin replied:
I'll open a separate issue but just on that point: A Sum represents an expression e.g. both sums above are in fact integers. A function is not anything like a number. The application of the function can give an expression but the function itself is a "callable". Also it's not necessarily the case that a Lambda even evaluates to an Expr:
```julia
In [22]: f = Lambda(x, (x, x))

In [23]: f(1)
Out[23]: (1, 1)
```

smichr replied:
Reading this now, I am guessing that you mean that Lambda addition should be prohibited like Eq addition.

smichr commented on sympy/core/tests/test_function.py:
```diff
@@ -216,10 +216,9 @@ def test_Lambda():
     assert Lambda((), 42).nargs == FiniteSet(0)
 
-    assert unchanged(Lambda, (x,), x**2)
+    assert Lambda(x, x**2).atoms(Symbol) == {Symbol('l')}
```
```suggestion
    assert unchanged(Lambda, (x,), x**2)
```

smichr commented on sympy/core/tests/test_function.py:
```diff
@@ -239,7 +238,9 @@ def test_Lambda():
 
-    assert Lambda(x, 2*x) + Lambda(y, 2*y) == 2*Lambda(x, 2*x)
+    eq = Lambda(x, 2*x) + Lambda(y, 2*y)
+    assert eq == 2*Lambda(x, 2*x)
```
```suggestion
    assert eq != 2*Lambda(x, 2*x)
```

smichr commented on sympy/functions/elementary/miscellaneous.py:
```diff
@@ -43,7 +43,7 @@ class IdentityFunction(Lambda, metaclass=Singleton):
 
     def __new__(cls):
-        x = Dummy('x')
+        x = Symbol('l')
```
```suggestion
        x = Dummy('x')
```

smichr commented on sympy/functions/elementary/miscellaneous.py:
```diff
@@ -11,7 +11,7 @@
 from sympy.core.relational import Eq, Relational
 from sympy.core.singleton import Singleton
-from sympy.core.symbol import Dummy
+from sympy.core.symbol import Symbol
```
```suggestion
from sympy.core.symbol import Dummy
```

smichr commented on sympy/printing/latex.py:
```diff
@@ -941,7 +941,7 @@ def _print_Lambda(self, expr):
 
     def _print_IdentityFunction(self, expr):
-        return r"\left( x \mapsto x \right)"
+        return r"\left( l \mapsto l \right)"
```
```suggestion
        return r"\left( x \mapsto x \right)"
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2242,23 +2242,23 @@ def test_pretty_product():
     expr = Lambda(y, y)
-    assert pretty(expr) == "x -> x"
-    assert upretty(expr) == u"x ↦ x"
+    assert pretty(expr) == "l -> l"
+    assert upretty(expr) == u"l ↦ l"
```
```suggestion
    assert pretty(expr) == "x -> x"
    assert upretty(expr) == u"x ↦ x"
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2242,23 +2242,23 @@ def test_pretty_product():
     expr = Lambda(x, x+1)
-    assert pretty(expr) == "x -> x + 1"
-    assert upretty(expr) == u"x ↦ x + 1"
+    assert pretty(expr) == "l -> l + 1"
+    assert upretty(expr) == u"l ↦ l + 1"
```
```suggestion
    assert pretty(expr) == "x -> x + 1"
    assert upretty(expr) == u"x ↦ x + 1"
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2242,23 +2242,23 @@ def test_pretty_product():
 """\
       2\n\
-x -> x \
+l -> l \
```
```suggestion
x -> x \
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2242,23 +2242,23 @@ def test_pretty_product():
 u("""\
      2\n\
-x ↦ x \
+l ↦ l \
```
```suggestion
x ↦ x \
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
          2
 /      2\\ \n\
-\\x -> x / \
+\\l -> l / \
```
```suggestion
\\x -> x / \
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
         2
 ⎛     2⎞ \n\
-⎝x ↦ x ⎠ \
+⎝l ↦ l ⎠ \
```
```suggestion
⎝x ↦ x ⎠ \
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
     expr = Lambda((x, y), x)
-    ascii_str = "(x, y) -> x"
-    ucode_str = u"(x, y) ↦ x"
+    ascii_str = "(l, l0) -> l"
+    ucode_str = u"(l, l₀) ↦ l"
```
```suggestion
    ascii_str = "(x, y) -> x"
    ucode_str = u"(x, y) ↦ x"
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
 """\
-           2\n\
-(x, y) -> x \
+            2\n\
+(l, l0) -> l \
```
```suggestion
           2\n\
(x, y) -> x \
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
 u("""\
-          2\n\
-(x, y) ↦ x \
+           2\n\
+(l, l₀) ↦ l \
```
```suggestion
          2\n\
(x, y) ↦ x \
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
 """)
-    assert pretty(expr) == ascii_str
-    assert upretty(expr) == ucode_str
+    assert pretty(expr) == ascii_str, pretty(expr)
+    assert upretty(expr) == ucode_str, upretty(expr)
```
```suggestion
    assert pretty(expr) == ascii_str
    assert upretty(expr) == ucode_str
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
 """\
-              2\n\
-((x, y),) -> x \
+               2\n\
+((l, l0),) -> l \
```
```suggestion
              2\n\
((x, y),) -> x \
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
 u("""\
-             2\n\
-((x, y),) ↦ x \
+              2\n\
+((l, l₀),) ↦ l \
```
```suggestion
             2\n\
((x, y),) ↦ x \
```

smichr commented on sympy/printing/pretty/tests/test_pretty.py:
```diff
@@ -2268,49 +2268,49 @@ def test_pretty_Lambda():
+((l, l₀),) ↦ l \
 """)
-    assert pretty(expr) == ascii_str
+    assert pretty(expr) == ascii_str, pretty(expr)
```
```suggestion
    assert pretty(expr) == ascii_str
```

smichr commented on sympy/printing/tests/test_latex.py:
```diff
@@ -1425,11 +1425,11 @@ def test_latex_order():
 def test_latex_Lambda():
     assert latex(Lambda(x, x + 1)) == \
-        r"\left( x \mapsto x + 1 \right)"
+        r"\left( l \mapsto l + 1 \right)"
```
```suggestion
        r"\left( x \mapsto x + 1 \right)"
```

smichr commented on sympy/printing/tests/test_latex.py:
```diff
@@ -1425,11 +1425,11 @@ def test_latex_order():
+        r"\left( l \mapsto l + 1 \right)"
     assert latex(Lambda((x, y), x + 1)) == \
-        r"\left( \left( x, \  y\right) \mapsto x + 1 \right)"
+        r"\left( \left( l, \  l_{0}\right) \mapsto l + 1 \right)"
```
```suggestion
        r"\left( \left( x, \  y\right) \mapsto x + 1 \right)"
```

smichr commented on sympy/printing/tests/test_latex.py:
```diff
@@ -1425,11 +1425,11 @@ def test_latex_order():
+        r"\left( \left( l, \  l_{0}\right) \mapsto l + 1 \right)"
     assert latex(Lambda(x, x)) == \
-        r"\left( x \mapsto x \right)"
+        r"\left( l \mapsto l \right)"
```
```suggestion
        r"\left( x \mapsto x \right)"
```

smichr commented on sympy/printing/tests/test_mathml.py:
```diff
@@ -1485,12 +1485,13 @@ def test_print_ceiling():
     expr = Lambda(x, x+1)
     assert mathml(expr, printer='presentation') == \
-        '<mfenced><mrow><mi>x</mi><mo>&#x21A6;</mo><mrow><mi>x</mi><mo>+</mo>'\
+        '<mfenced><mrow><mi>l</mi><mo>&#x21A6;</mo><mrow><mi>l</mi><mo>+</mo>'\
```
```suggestion
        '<mfenced><mrow><mi>x</mi><mo>&#x21A6;</mo><mrow><mi>x</mi><mo>+</mo>'\
```

smichr commented on sympy/printing/tests/test_mathml.py:
```diff
@@ -1485,12 +1485,13 @@ def test_print_ceiling():
     assert mathml(expr, printer='presentation') == \
-        '<mfenced><mrow><mrow><mfenced><mi>x</mi><mi>y</mi></mfenced></mrow>'\
-        '<mo>&#x21A6;</mo><mrow><mi>x</mi><mo>+</mo><mi>y</mi></mrow></mrow></mfenced>'
+        '<mfenced><mrow><mrow><mfenced><mi>l</mi><msub><mi>l</mi><mi>0</mi>'\
+        '</msub></mfenced></mrow><mo>&#x21A6;</mo><mrow><mi>l</mi><mo>+</mo>'\
+        '<msub><mi>l</mi><mi>0</mi></msub></mrow></mrow></mfenced>'
```
```suggestion
        '<mfenced><mrow><mrow><mfenced><mi>x</mi><mi>y</mi></mfenced></mrow>'\
        '<mo>&#x21A6;</mo><mrow><mi>x</mi><mo>+</mo><mi>y</mi></mrow></mrow></mfenced>'
```

smichr commented on sympy/printing/tests/test_str.py:
```diff
@@ -180,12 +180,12 @@ def test_AccumBounds():
 
 def test_Lambda():
-    assert str(Lambda(d, d**2)) == "Lambda(_d, _d**2)"
+    assert str(Lambda(d, d**2)) == "Lambda(l, l**2)"
```
```suggestion
    assert str(Lambda(d, d**2)) == "Lambda(_d, _d**2)"
```

smichr commented on sympy/printing/tests/test_str.py:
```diff
@@ -180,12 +180,12 @@ def test_AccumBounds():
     assert str(Lambda((), x)) == "Lambda((), x)"
-    assert str(Lambda((x, y), x+y)) == "Lambda((x, y), x + y)"
-    assert str(Lambda(((x, y),), x+y)) == "Lambda(((x, y),), x + y)"
+    assert str(Lambda((x, y), x + y)) == "Lambda((l, l0), l + l0)"
+    assert str(Lambda(((x, y),), x + y)) == "Lambda(((l, l0),), l + l0)"
```
```suggestion
    assert str(Lambda((x, y), x + y)) == "Lambda((x, y), x + y)"
    assert str(Lambda(((x, y),), x + y)) == "Lambda(((x, y),), x + y)"
```

smichr commented on sympy/core/basic.py:
```diff
@@ -941,10 +937,14 @@ def subs(self, *args, **kwargs):
+            # \___ 2 __/    \_1_/  <- number of nodes
+            # the normal keys are applied first and then the
+            # ordering by _nodes is applied
+            k = list(ordered(sequence, keys=lambda x: -_nodes(x))
+            sequence = [(k, sequence[k]) for k in k]
```
```suggestion
            sequence = list(ordered(sequence, keys=lambda x: -_nodes(x)))
```

smichr commented on sympy/core/basic.py:
```diff
@@ -941,10 +937,13 @@ def subs(self, *args, **kwargs):
+            # \___ 2 __/    \_1_/  <- number of nodes
+            # the normal keys are applied first and then the
+            # ordering by _nodes is applied
+            sequence = list(ordered(sequence, keys=lambda x: -_nodes(x)))
```
```suggestion
            k= list(ordered(sequence, keys=lambda x: -_nodes(x)))
            sequence = [(k, sequence[k]) for k in k]
```

smichr committed:
19326: ordering subs dict

To make substitution canonical, unordered
replacements must be ordered in some way.
Although more complex topological sorting
is possible, the current method is to put
more complex patterns first and for those
that have the same complexity, order them
by using the default_sort_key.

This commit corrects the original change
which put more complex patterns first but
did not account for the keeping order amongst
the patterns that had the same complexity.

The _nodes function had to be modified to put
higher order derivatives ahead of those with
lower order.

smichr committed:
remove custom Lambda.__eq__

Python `==` is meant to tell when two objects can be used
interchangeably. When users write expressions in different
variables, SymPy does not make those variables canonical so
such expressions must not compare as equal.

smichr committed:
make replace more robust

Not every object in a tree can be replaced with a Dummy symbol
`replace` no longer does this.

smichr committed:
use dummy_eq in tests

smichr committed:
match ignores bound symbols unless requested

This means that when matching expressions with
potentially different bound symbols, if you
aren't interested in the identity of the
bound symbols you can leave them out of the
pattern.

smichr committed:
sort match terms

When matching, args are ordered but the default sort key
prioritizes symbols over leading Numbers on Mul. This means
5*x < 2*y. By sorting by the leading Number (as the added
test shows) a better match can be made. Previously the test
gave

>>> (5*y-x).match(5*Wild('x')-Wild('y'))
{Y_: -5*y, X_: -x/5}

Now it gives

>>> (5*y-x).match(5*Wild('x')-Wild('y'))
{Y_: x, X_: y}

smichr committed:
use == not is for comparison

smichr commented:
Wow -- would not have expected this and haven't tracked down how it is happening, but an extra symbol is finding its way into the match result.
```
pat, expr = Sum(CG(a_, _0, b_, 0, a_, _0), (_0, -a_, a_)), Sum(CG(a, _0, b, 0, a, _0), (_0, -a, a))
pat.match(expr) -> {V_: x, a_: a, b_: b}
```
Where is that `V_` coming from?? It doesn't happen under 3.7 (Windows) but is happening under 3.8 during travis tests. I am now sanitizing the match dictionary to correct for this. I imagine that it gets in when the dictionary is treated like a defaultdict and a query like `d[V_]` is made:
```
>>> d
defaultdict(<class 'int'>, {})
>>> d[x]
0
>>> d
defaultdict(<class 'int'>, {x: 0})
```
But how it ends up giving an `x` for the value...must be a special defaultdict somewhere...

smichr commented:
@oscarbenjamin , every issue has been resolved here. The injection of the "V_" into the match results -- a system/Python-dependent incident -- has been worked around. I don't know the root cause of that, however, as I mention [above](https://github.com/sympy/sympy/pull/19261#issuecomment-630589553).

smichr commented:
@iammosespaulr , if you could take a look through these many small changes that would be helpful.

smichr commented:
@iammosespaulr , each commit is self contained if that helps you work through this.

oscarbenjamin commented:
All looks good except not sure about the last commit. If I go back one commit from the tip of this branch how would I potentially reproduce the problem?

smichr commented:
> how would I potentially reproduce the problem?

push the commit to travis - the first assertion in `test_cg_simp_sum` will fail.  The helper for the simplifcation routine checks the length of the match dictionary and, because a spurious symbol formerly came back in the dictionary, the assumption that only the requested symbols would be present is violated. It failed every time on travis after the `302bdfd` commit ("match ignores bound symbols unless requested"); it never fails locally under Windows with Python 3.7.

smichr commented:
@oscarbenjamin , I moved the test that will fail to `basic` and put an assertion in the match function. It should fail on my `debug` branch travis [run](https://travis-ci.org/github/smichr/sympy/builds/688918208?utm_source=github_status&utm_medium=notification).

oscarbenjamin commented:
Looks like lots of things are failing on the debug branch

smichr commented:
> debug branch

https://travis-ci.org/github/smichr/sympy/builds/688918208

iammosespaulr commented:
> @iammosespaulr , each commit is self contained if that helps you work through this.

It's kinda a big pr, I'm starting from the issues fixed 😄, also checking out your `debug` branch.

smichr commented:
> Looks like lots of things are failing on the debug branch

The assertion that there be no extraneous symbols in the match is what is causing others to fail. But the code using those results wasn't influenced by the presence of extra symbols.

VERY CURIOUS that the same test, moved to basic, did not fail but failed in the cg context. I somewhat think this is a Python issue and not a SymPy issue (directly).

smichr commented:
> also checking out your `debug` branch.

My humble recommendation is to not worry about that branch. I'm not sure that it's really a SymPy issue. Going through the commits as they appear here in order -- each is pretty small -- would be a good way to go. But your choice. :-)

oscarbenjamin commented:
> VERY CURIOUS that the same test, moved to basic, did not fail but failed in the cg context.

I think it's because it depends on one test having been run before another.

smichr commented:
> I think it's because it depends on one test having been run before another.

Is there some issue that we are abusing by providing a default argument to the match routine:

`def matches(...repl_dict={}...)` Should that be None and then internally provide a dictionary if it is None?

oscarbenjamin commented:
> Is there some issue that we are abusing by providing a default argument to the match routine:
> 
> `def matches(...repl_dict={}...)` Should that be None and then internally provide a dictionary if it is None?

Oh, yes. That could be the cause of this. It should be fixed either way. Don't use mutable defaults!

The dict is directly returned if `self==expr`:
https://github.com/sympy/sympy/blob/0180df08428d4c9ccff124fd2984720e8409f46f/sympy/core/basic.py#L1556-L1589
That allows the caller to modify it which will affect subsequent calls to matches.

smichr commented:
> Don't use mutable defaults!

this should be a TODO -- there are 70 cases of `={}` in the codebase and 71 of `=[]`.

oscarbenjamin commented:
Here's a demo:
```julia
In [1]: exp(x).matches(exp(y))                                                                                                    

In [2]: d = exp(x).matches(exp(x))                                                                                                

In [3]: d[x] = y                                                                                                                  

In [4]: exp(x).matches(exp(y))                                                                                                    
Out[4]: {x: y}
```

smichr commented:
So should None be used or should we just make sure that each routine that uses it starts with `repl_dict = repl_dict.copy()`?

smichr committed:
always use copy of repl_dict

smichr commented:
@oscarbenjamin , assuming you don't want this in 1.6, I will look to merge as soon as 1.6 is released.

oscarbenjamin commented:
1.6 is already on a separate branch so this can go into master.

I think we should always use None to avoid mutable defaults so that there is no confusion. Then the rule is simple: no mutable defaults.

smichr commented:
@iammosespaulr, Oscar has already looked this over and the mystery issue has been solved. I'll wait to see if you have any comments and, if not, will commit in about 24 hours.

oscarbenjamin commented:
Yes, this looks good. I think it's better to use None as the default value rather than calling copy but that can be fixed in a more general cleanup for #19372

iammosespaulr commented on sympy/core/basic.py:
```diff
@@ -1480,55 +1487,40 @@ def replace(self, query, value, map=False, simultaneous=True, exact=None):
+            return rv
+
+
+        mapping = []  # changes that took place
```
```suggestion
        mapping = {}  # changes that took place
```

iammosespaulr commented on sympy/core/basic.py:
```diff
@@ -1480,55 +1487,40 @@ def replace(self, query, value, map=False, simultaneous=True, exact=None):
+                v = _value(expr, result)
+                if v is not None and v != expr:
+                    if map:
+                        mapping.append((expr, v))
```
```suggestion
                        mapping[expr] = v
```

iammosespaulr commented on sympy/core/basic.py:
```diff
@@ -1480,55 +1487,40 @@ def replace(self, query, value, map=False, simultaneous=True, exact=None):
-                        for k, v in mapping.items()}
-            return rv, mapping
+        rv = walk(self, rec_replace)
+        return (rv, dict(mapping)) if map else rv
```
```suggestion
        return (rv, mapping) if map else rv
```

smichr replied:
At one point I was using the ordered replacements to create RV. You are right, this can go back to dict.

iammosespaulr commented:
@smichr saw that a list was used and then converted back to dict, it was in `core/basic` so I thought one less step makes it faster right ?

iammosespaulr commented:
( also you've made extensive use of `dumeq`, does it enhance `dummy_eq`? Would it be useful to have it's functionality inside `dummy_eq` so that you don't have to declare a seperate func ? ) just a thought

smichr commented:
dumeq just handles the iterable case. Nonbasic list and tuple can't use `dummy_eq` method.

oscarbenjamin commented:
> dumeq just handles the iterable case. Nonbasic list and tuple can't use `dummy_eq` method.

This is one reason why I think it would be better if we used functions more rather than methods e.g. we could have:
```
>>> subs([x+y, x-y], x, 2)
[2+y, 2-y]
```
Same for evalf etc
